packages/core/diagnostic/src/diagnostic.js

```diff
@@ -0,0 +1,99 @@
+// @flow
+
+export type DiagnosticHighlightLocation = {|
+  line: number,
+  column: number
+|};
+
+export type DiagnosticSeverity = 'error' | 'warn' | 'info';
+
+export type DiagnosticCodeHighlight = {|
+  start: DiagnosticHighlightLocation,
+  end: DiagnosticHighlightLocation,
+  message?: string
+|};
+
+export type DiagnosticCodeFrame = {|
+  code: string,
+  codeHighlights: DiagnosticCodeHighlight | Array<DiagnosticCodeHighlight>
+|};
+
+// A Diagnostic is a style agnostic way of emitting errors, warnings and info
+// The reporter's are responsible for rendering the message, codeframes, hints, ...
+export type Diagnostic = {|
+  message: string,
+  //? severity: DiagnosticSeverity, // Might add this back later...
+  origin: string, // Name of plugin or file that threw this error
+
+  // Asset metadata
+  filename?: string,
+  language?: string,
+
+  // Codeframe data
+  codeframe?: DiagnosticCodeFrame,
+
+  // Hints to resolve issues faster
+  hints?: string | Array<string>,
+
+  //! Should only be used if there's no way to supply code and codeHighlight...
+  stack?: string
+|};
+
+// This type should represent all error formats Parcel can encounter...
+type PrintableError = Error & {
+  fileName: ?string,
+  codeFrame?: string,
+  highlightedCodeFrame?: string,
+  loc?: {
+    column: number,
+    line: number,
+    ...
+  },
+  source?: string,
+  ...
+};
+
+export function errorToDiagnostic(error: PrintableError): Diagnostic {
```

Comment:
Should this fetch source from FileSystem if `error.fileName` isn't undefined and `error.source` is undefined.